### PR TITLE
action: support for DRA

### PR DESCRIPTION
**Pull request by v1v**
## Motivation/summary

Use Buildkite to package and run the DRA.

This will run on push events

## How to test these changes

In the CI

## Related issues

Requires https://github.com/elastic/apm-server/pull/11373 and https://github.com/elastic/apm-server/pull/11399

## Tasks

- [x] exclude if main and staging.
- [x] exclude if branch is not supported anymore isBranchUnifiedReleaseAvailable()
- [x] upload/download artifacts so they can be consumed when running the DRA.
- [x] Add pipeline in the catalog entry

## Tests

It worked like a charm using a `dry-run` with some ignored steps regarding running on a PR or a non Unified Release branch.

This [buildkite build](https://buildkite.com/elastic/apm-server-package/builds/41#018a04cc-bf3a-409e-bb91-9efef0814057) is the one that was executed with dry-run based on the ignored validations in https://github.com/elastic/apm-server/pull/11400/commits/22dcf814aa63e750b1415cad136a18aad2a470c5

image

image

### CODE REVIEW

File: .buildkite/scripts/package.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+if [[ ${PLATFORM_TYPE} == "arm" || ${PLATFORM_TYPE} == "aarch64" ]]; then
+  MAKE_GOAL=package-docker
+fi
+
+if [[ ${TYPE} == "snapshot" ]]; then
+  MAKE_GOAL="${MAKE_GOAL}-snapshot"
+fi
+
+echo "--- Configure golang :golang:"
+eval "$(./gvm $(cat .go-version))"
```
Comment (v1v) [inline .buildkite/scripts/package.sh:23]: Ideally the PATH should updated but for some reason `.buildkite/hooks/pre-command` does not export env variables accordingly.

For instance `GO_VERSION` is exported , see https://github.com/elastic/apm-server/pull/11400/files#diff-71e4f7cc0d357183cbbb900bb97a29db006263fb1967d06035e417c38f27e53dR80, but it cannot be used with something like

```suggestion
eval "$(./gvm $GO_VERSION)"
```

File: .github/workflows/package.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      #   name: Run buildkite pipeline
+      #   # NOTE: exclude if main and staging
+      #   if: ${{ matrix.dra-workflow != 'staging' && github.ref != 'refs/heads/main' }}
+      #   uses: elastic/apm-pipeline-library/.github/actions/buildkite@current
+      #   with:
+      #     vaultUrl: ${{ secrets.VAULT_ADDR }}
+      #     vaultRoleId: ${{ secrets.VAULT_ROLE_ID }}
+      #     vaultSecretId: ${{ secrets.VAULT_SECRET_ID }}
+      #     #pipeline: apm-server-package
+      #     pipeline: apm-server
```
Comment (v1v) [inline .github/workflows/package.yml:49]:
```suggestion
      #     pipeline: apm-server-package
```

File: .github/workflows/opentelemetry.yml [hunk 3]
```diff
@@ -7,6 +7,7 @@ on:
       - bump-golang
       - ci
       - microbenchmark
+      - package
```
Comment (v1v) [inline .github/workflows/opentelemetry.yml:10]:
```suggestion
```

File: .buildkite/hooks/pre-command [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+PATH="${PATH}:$GOPATH:$GOPATH/bin"
+export PATH
+
+echo "--- Prepare vault context :vault:"
+#CI_DRA_ROLE_PATH="kv/ci-shared/release/dra-role"
+#DRA_CREDS_SECRET=$(retry 5 vault kv get -field=data -format=json ${CI_DRA_ROLE_PATH})
+#VAULT_ADDR_SECRET=$(echo ${DRA_CREDS_SECRET} | jq -r '.vault_addr')
+#VAULT_ROLE_ID_SECRET=$(echo ${DRA_CREDS_SECRET} | jq -r '.role_id')
+#VAULT_SECRET=$(echo ${DRA_CREDS_SECRET} | jq -r '.secret_id')
+#export VAULT_ADDR_SECRET VAULT_ROLE_ID_SECRET VAULT_SECRET
```
Comment (v1v) [inline .buildkite/hooks/pre-command:99]:
```suggestion
CI_DRA_ROLE_PATH="kv/ci-shared/release/dra-role"
DRA_CREDS_SECRET=$(retry 5 vault kv get -field=data -format=json ${CI_DRA_ROLE_PATH})
VAULT_ADDR_SECRET=$(echo ${DRA_CREDS_SECRET} | jq -r '.vault_addr')
VAULT_ROLE_ID_SECRET=$(echo ${DRA_CREDS_SECRET} | jq -r '.role_id')
VAULT_SECRET=$(echo ${DRA_CREDS_SECRET} | jq -r '.secret_id')
export VAULT_ADDR_SECRET VAULT_ROLE_ID_SECRET VAULT_SECRET
```

File: .buildkite/package.yml [hunk 5]
```diff
@@ -1,5 +1,55 @@
-# yaml-language-server: $schema=https://raw.githubusercontent.com/buildkite/pipeline-schema/main/schema.json
+notify:
+  - slack:
+      channels:
+        #- "#apm-server"
+        - "#observablt-bots"
```
Comment (v1v) [inline .buildkite/package.yml:5]:
```suggestion
        - "#apm-server"
```

File: .buildkite/hooks/pre-command [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+if command -v docker &>/dev/null; then
+    echo "--- Prepare docker context :docker:"
+    DOCKER_REGISTRY_SECRET_PATH="secret/ci/elastic-apm-server/docker-elastic-observability"
+    DOCKER_USERNAME_SECRET=$(retry 5 vault kv get -field username "${DOCKER_REGISTRY_SECRET_PATH}")
+    DOCKER_PASSWORD_SECRET=$(retry 5 vault kv get -field password "${DOCKER_REGISTRY_SECRET_PATH}")
+    DOCKER_REGISTRY_SECRET=$(retry 5 vault kv get -field registry "${DOCKER_REGISTRY_SECRET_PATH}")
+    docker login -u "${DOCKER_USERNAME_SECRET}" -p "${DOCKER_PASSWORD_SECRET}" "${DOCKER_REGISTRY_SECRET}"
+    unset DOCKER_USERNAME_SECRET DOCKER_PASSWORD_SECRET
+    export DOCKER_REGISTRY_SECRET
+	retry 4 docker pull --quiet docker.elastic.co/infra/release-manager:latest
```
Comment (v1v) [inline .buildkite/hooks/pre-command:76]:
```suggestion
    retry 4 docker pull --quiet docker.elastic.co/infra/release-manager:latest
```

File: .buildkite/scripts/dra.sh [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+ls -l build/
+
+if [[ "${BUILDKITE_PULL_REQUEST:-false}" == "false" ]]; then
+  echo "Release Manager does not run on PRs, skipping"
+  exit 0
+fi
+
+curl -s https://storage.googleapis.com/artifacts-api/snapshots/branches.json > active-branches.json
+if ! grep -q "\"$BUILDKITE_BRANCH\"" active-branches.json ; then
+  echo "Release Manager only support the current active branches, skipping"
```
Comment (v1v) [inline .buildkite/scripts/dra.sh:34]:
```suggestion
  echo "Release Manager only supports the current active branches, skipping"
```

File: .buildkite/package.yml [hunk 8]
```diff
@@ -1,5 +1,55 @@
-# yaml-language-server: $schema=https://raw.githubusercontent.com/buildkite/pipeline-schema/main/schema.json
+notify:
+  - slack:
+      channels:
```
Comment (v1v) [inline .buildkite/package.yml:3]:
```suggestion
    if: build.state == "failed"
      channels:
```

File: .buildkite/package.yml [hunk 9]
```diff
@@ -1,5 +1,56 @@
-# yaml-language-server: $schema=https://raw.githubusercontent.com/buildkite/pipeline-schema/main/schema.json
+notify:
+  - slack:
+    if: build.state == "failed"
+      channels:
+        #- "#apm-server"
+        - "#observablt-bots"
```
Comment (v1v) [inline .buildkite/package.yml:6]:
```suggestion
notify:
  #- slack: "#apm-server"
  - slack: "#observablt-bots"
    if: build.state == "failed"
```

File: .buildkite/scripts/dra.sh [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+buildkite-agent artifact download "build/dependencies*.csv" .
+
+echo "--- Changing permissions for the release manager"
+sudo chown -R :1000 build/
+ls -l build/distributions/
+ls -l build/
+
+if [[ "${BUILDKITE_PULL_REQUEST:-false}" == "false" ]]; then
+  echo "--- :arrow_right: Release Manager does not run on PRs, skipping"
+  exit 0
```
Comment (v1v) [inline .buildkite/scripts/dra.sh:27]:
```suggestion
  # TODO: for testing purposes
  #exit 0
```

File: .buildkite/scripts/dra.sh [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+curl -s https://storage.googleapis.com/artifacts-api/snapshots/branches.json > active-branches.json
+if ! grep -q "\"$BUILDKITE_BRANCH\"" active-branches.json ; then
+  echo "--- :arrow_right: Release Manager only supports the current active branches, skipping"
+  echo "BUILDKITE_BRANCH=$BUILDKITE_BRANCH"
+  echo "BUILDKITE_COMMIT=$BUILDKITE_COMMIT"
+  echo "VERSION=$VERSION"
+  echo "Supported branches:"
+  cat active-branches.json
+  exit 0
```
Comment (v1v) [inline .buildkite/scripts/dra.sh:38]:
```suggestion
  # TODO: for testing purposes
  #exit 0
```

File: .buildkite/scripts/dra.sh [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    -e VAULT_SECRET_ID="${VAULT_SECRET}" \
+    --mount type=bind,readonly=false,src=$(pwd),target=/artifacts \
+    docker.elastic.co/infra/release-manager:latest \
+      cli collect \
+      --project apm-server \
+      --branch $BUILDKITE_BRANCH \
+      --commit $BUILDKITE_COMMIT \
+      --workflow $workflow \
+      --artifact-set main \
+      --version $VERSION
```
Comment (v1v) [inline .buildkite/scripts/dra.sh:57]:
```suggestion
      --version $VERSION \
      --dry-run
      ## TODO: for testing purposes it uses --dry-run
```

File: .buildkite/scripts/dra.sh [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+  echo "VERSION=$VERSION"
+  echo "Supported branches:"
+  cat active-branches.json
+  exit 0
+fi
+
+dra() {
+  local workflow=$1
+  echo "--- Run release manager $workflow"
+  docker run --rm \
```
Comment (v1v) [inline .buildkite/scripts/dra.sh:44]:
```suggestion
  # TODO: for testing purposes let's force the branch
  BUILDKITE_BRANCH=main
  docker run --rm \
```

File: .buildkite/scripts/dra.sh [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+VERSION=$(make get-version)
+
+echo "--- Restoring Artifacts"
+buildkite-agent artifact download "build/**/*" .
+buildkite-agent artifact download "build/dependencies*.csv" .
+
+echo "--- Changing permissions for the release manager"
+sudo chown -R :1000 build/
+ls -l build/distributions/
+ls -l build/
```
Comment (v1v) [inline .buildkite/scripts/dra.sh:23]:
```suggestion
```

File: .buildkite/scripts/dra.sh [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+##
+
+set -eo pipefail
+
+## Read current version.
+VERSION=$(make get-version)
+
+echo "--- Restoring Artifacts"
+buildkite-agent artifact download "build/**/*" .
+buildkite-agent artifact download "build/dependencies*.csv" .
```
Comment (v1v) [inline .buildkite/scripts/dra.sh:18]:
```suggestion
buildkite-agent artifact download "build/dependencies*.csv" .
# The dependencies file needs to be saved in the build/distributions folder
cp build/dependencies*.csv build/distributions/
```

File: .buildkite/scripts/dra.sh [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+curl -s https://storage.googleapis.com/artifacts-api/snapshots/branches.json > active-branches.json
+if ! grep -q "\"$BUILDKITE_BRANCH\"" active-branches.json ; then
+  echo "--- :arrow_right: Release Manager only supports the current active branches, skipping"
+  echo "BUILDKITE_BRANCH=$BUILDKITE_BRANCH"
+  echo "BUILDKITE_COMMIT=$BUILDKITE_COMMIT"
+  echo "VERSION=$VERSION"
+  echo "Supported branches:"
+  cat active-branches.json
+  # TODO: for testing purposes
+  #exit 0
```
Comment (v1v) [inline .buildkite/scripts/dra.sh:41]:
```suggestion
  exit 0
```

File: .buildkite/scripts/dra.sh [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+cp build/dependencies*.csv build/distributions/
+
+echo "--- Changing permissions for the release manager"
+sudo chown -R :1000 build/
+ls -l build/distributions/
+
+if [[ "${BUILDKITE_PULL_REQUEST:-false}" == "false" ]]; then
+  echo "--- :arrow_right: Release Manager does not run on PRs, skipping"
+  # TODO: for testing purposes
+  #exit 0
```
Comment (v1v) [inline .buildkite/scripts/dra.sh:29]:
```suggestion
  exit 0
```

File: .buildkite/scripts/dra.sh [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+    docker.elastic.co/infra/release-manager:latest \
+      cli collect \
+      --project apm-server \
+      --branch $BUILDKITE_BRANCH \
+      --commit $BUILDKITE_COMMIT \
+      --workflow $workflow \
+      --artifact-set main \
+      --version $VERSION \
+      --dry-run
+      ## TODO: for testing purposes it uses --dry-run
```
Comment (v1v) [inline .buildkite/scripts/dra.sh:64]:
```suggestion
      --version $VERSION
```

File: .buildkite/package.yml [hunk 19]
```diff
@@ -1,5 +1,54 @@
-# yaml-language-server: $schema=https://raw.githubusercontent.com/buildkite/pipeline-schema/main/schema.json
+notify:
+  #- slack: "#apm-server"
+  - slack: "#observablt-bots"
```
Comment (v1v) [inline .buildkite/package.yml:3]:
```suggestion
  - slack: "#apm-server"
```

File: .buildkite/scripts/dra.sh [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+  echo "Supported branches:"
+  cat active-branches.json
+  exit 0
+fi
+
+dra() {
+  local workflow=$1
+  echo "--- Run release manager $workflow"
+  # TODO: for testing purposes let's force the branch
+  BUILDKITE_BRANCH=main
```
Comment (v1v) [inline .buildkite/scripts/dra.sh:46]:
```suggestion
```

**Review by axw — APPROVED** [review]

**Review by amannocci — APPROVED** [review]

**Pull request merged**